Copy to option

priyanshu16095 opened the pull request:
### Fixes #12339, #12340 and #12341

This PR introduces a 'Copy to' context menu option with features for cross-reference inclusion/exclusion, as well as the ability to save user preferences.

### THE CHANGES MADE
- Implemented a 'Copy to' context menu that lists the names of currently opened libraries, excluding the active one, as submenu items. Selecting a library from the submenu copies the selected entry to the chosen library.
- If an entry includes a cross-referenced entry, a dialog box is displayed, allowing users to choose whether to include or exclude the cross-references.
- Added functionality to save user preferences for cross-reference inclusion/exclusion, ensuring a personalized experience for future use.

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

![Screenshot (280)](https://github.com/user-attachments/assets/d278acba-5cf6-41ce-b566-cb789c380991)
![b](https://github.com/user-attachments/assets/12fa646d-416b-43a2-9716-0d2ae1515364)

priyanshu16095 commented:
Please review it and provide any suggestions.

Siedlerchr commented:
We will do, please keep in mind that all maintainers work at jabref in
their leisure time despite having a full time job. I will try to look into
it tomorrow.

Priyanshu Gupta ***@***.***> schrieb am Sa., 11. Jan. 2025,
20:38:

> Please review it and provide any suggestions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/12374#issuecomment-2585386144>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACOFZA3TI5CFOYLSVHGPHL2KFXKDAVCNFSM6AAAAABU5GXCISVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBVGM4DMMJUGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

priyanshu16095 commented:
Apologies, and thanks for the update! I understand, and I really appreciate your time.

subhramit commented on src/main/java/org/jabref/gui/edit/CopyTo.java:
```diff
@@ -0,0 +1,94 @@
+    public void execute() {
+        Logger logger = LoggerFactory.getLogger(CopyTo.class);
+
+        BibDatabaseContext databaseContext = new BibDatabaseContext();
```
Not needed here.

priyanshu16095 replied:
Sure, I will remove them, considering it is a draft PR. I left them for explanatory purposes.

subhramit commented on src/main/java/org/jabref/gui/edit/CopyTo.java:
```diff
@@ -0,0 +1,94 @@
+                .map(entry -> entry.getTitle().get())
+                .toList();
+
+        for (String title: titles) {
+            logger.info("Selected Entreis: " + title);
+        }
+        for (String checkedPath: checkedPaths) {
+            logger.info("Selected Libraries to copy: " + checkedPath);
+        }
```
Remove (loops with no utility apart from logging)

Siedlerchr replied:
Logic for copying the cross ref stuff is missing But I think @HoussemNasri  splitted that task?

HoussemNasri replied:
No, the copying of cross-ref stuff is the ticket assigned to @priyanshu16095.

https://github.com/JabRef/jabref/issues/12339 is the ticket for creating the "copy to" menu; it is assigned to someone else, but since @priyanshu16095 already started working on it we agreed to have a draft pull request that might eventually be considered for merge in case of inactivity from the current assignee of that issue.

The issues/tasks are tightly coupled and meant to be done in a certain order. This pull request might or might not be merged depending on whether we receive any activities from the assignee of #12339, but IMO the work done here could be reused later to tackle #12341.

HoussemNasri replied:
> Logic for copying the cross ref stuff is missing But I think @HoussemNasri splitted that task?

Actually, #12340 was meant to be tackled before the one mentioned in PR description https://github.com/JabRef/jabref/issues/12341, so yes logic for copy cross references is for another task or could be handled in this PR since both issues are assigned to @priyanshu16095.

subhramit commented on src/main/java/org/jabref/gui/edit/CopyTo.java:
```diff
@@ -0,0 +1,94 @@
+        boolean includeCrossReferences = askForCrossReferencedEntries();
+        preferences.getCopyToPreferences().setShouldIncludeCrossReferences(includeCrossReferences);
+
+        if (preferences.getCopyToPreferences().getShouldIncludeCrossReferences()) {
+            logger.info("Include Cross References");
+        } else {
+            logger.info("Exclude Cross References");
+        }
```
Either decrease log noise or remove.

subhramit commented on src/main/java/org/jabref/gui/edit/CopyTo.java:
```diff
@@ -0,0 +1,94 @@
+                    optOut -> preferences.getCopyToPreferences().setShouldAskForIncludingCrossReferences(!optOut)
+            );
+        } else {
+            return preferences.getCopyToPreferences().getShouldIncludeCrossReferences();
```
See if you can improve abstraction - pass `preferences.getCopyToPreferences()` via constructor instead of entire `GuiPreferences`

subhramit commented on src/main/java/org/jabref/gui/maintable/RightClickMenu.java:
```diff
@@ -103,6 +110,37 @@ public static ContextMenu create(BibEntryTableViewModel entry,
+        if (!openDatabases.isEmpty()) {
+            openDatabases.forEach(library -> {
+                String path = library.getDatabasePath().get().toString();
+                String name = path.substring(path.lastIndexOf('\\') + 1);
```
@calixtus i think we use [`FromAuxDialogViewModel#getDatabaseName()`](https://github.com/JabRef/jabref/blob/2cd83dfdb056482aad9595489a4a7c4b796f6e37/src/main/java/org/jabref/gui/auximport/FromAuxDialogViewModel.java#L69) here?

Siedlerchr replied:
Yeah we also use that code part in
https://github.com/JabRef/jabref/blob/7c958e7dd067adbfd7244224aa551c6b3f1e9721/src/main/java/org/jabref/gui/importer/ImportEntriesDialog.java#L122-L130 
We should add this somewhere more central

subhramit reviewed (changes requested):
Comments on first look.

subhramit commented:
Please also fix the Checkstyle, Openrewrite and Unit tests. The [FAQ page](https://devdocs.jabref.org/code-howtos/faq.html) will help you.

HoussemNasri commented on src/main/java/org/jabref/gui/edit/CopyToPreferences.java:
```diff
@@ -0,0 +1,32 @@
+        this.shouldAskForIncludingCrossReferences.set(shouldAskForIncludingCrossReferences);
+    }
+
+    public boolean getShouldIncludeCrossReferences() {
```
How about `shouldIncludeCrossReferences`? It reads better.

priyanshu16095 replied:
Is it in line 17 or 14?

HoussemNasri replied:
The method name. It is a nitpick so feel free to ignore.

HoussemNasri commented on src/main/java/org/jabref/gui/edit/CopyTo.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Override
+    public void execute() {
+        Logger logger = LoggerFactory.getLogger(CopyTo.class);
```
Each class should declare the logger only once. Refer to the logging usage in other classes for guidance.

priyanshu16095 replied:
Oh, I used it for temporary purposes but forgot to remove it.

priyanshu16095 commented:
Sure, I left that logger code in place to help the reviewer understand it better.

subhramit commented:
> What should I do next? Is this PR now eligible for merging?

You can refer to Houssem's replies and ask him for further steps if any: https://github.com/JabRef/jabref/pull/12374#discussion_r1912538894

priyanshu16095 commented:
@HoussemNasri I’ve already included the code for #12340 in the pull request.
This task involves showing a dialog to provide a choice for including or excluding cross-references.

priyanshu16095 commented:
> @HoussemNasri I’ve already included the code for https://github.com/JabRef/jabref/issues/12340 in the pull request.  
> This task involves showing a dialog to provide a choice for including or excluding cross-references.
>
> This PR implements the following tasks:
>
> https://github.com/JabRef/jabref/issues/12339: Implemented a 'Copy to' context menu that displays the names of currently opened libraries as checkable menu items.  
> https://github.com/JabRef/jabref/issues/12340: Added a dialog that appears after selecting the libraries, allowing users to include or exclude cross-references.  
> https://github.com/JabRef/jabref/issues/12341: Implemented functionality to save user preferences for cross-reference inclusion or exclusion, ensuring a personalized experience for future use.

Have I understood correctly?  
Is there any feature left to implement in this PR?

HoussemNasri commented:
@priyanshu16095 why are the menu items checkable? We need to copy the bib entry to the library that the user chose in the menu. Also, I see no code that does the copying (neither normal copying nor cross-ref copying). Please re-read the issue descriptions.

priyanshu16095 commented:
Yeah, my bad. I thought it was only for creating the UI.

priyanshu16095 commented:
https://github.com/user-attachments/assets/ab6a43fc-4e7c-4300-8c01-5aeb028e7a8b

The feature is working fine; it copies the entry to the selected library.

If an entry has a cross-referenced entry, it shows dialog, allowing the user to choose whether to include or exclude the cross-referenced entry.

priyanshu16095 commented:
![Screenshot (278)](https://github.com/user-attachments/assets/5bda3c73-cf22-46e3-8f02-2d32a1ddc9e1)

Copying a duplicate entry into the library is causing this error.

subhramit commented on src/main/java/org/jabref/gui/edit/CopyToPreferences.java:
```diff
@@ -0,0 +1,32 @@
+    private final BooleanProperty shouldIncludeCrossReferences = new SimpleBooleanProperty();
+    private final BooleanProperty shouldAskForIncludingCrossReferences = new SimpleBooleanProperty();
+
+    public CopyToPreferences(boolean shouldAskForIncludingCrossReferences,
+                             boolean shouldIncludeCrossReferences
+                             ) {
```
Nitpick - make it single line (as only two parameters)

subhramit commented on src/main/java/org/jabref/gui/maintable/RightClickMenu.java:
```diff
@@ -103,6 +115,51 @@ public static ContextMenu create(BibEntryTableViewModel entry,
+        if (!openDatabases.isEmpty()) {
+            openDatabases.forEach(bibDatabaseContext -> {
+                Optional<Path> destinationPath = Optional.empty();
+                String destinationDatabaseName = " ";
```
Tip: Don't initialize with any non-default value that you will not use (in this case, a space). If you really want to, you can assign empty string (`""`). Preferably, don't assign (java will initialize as null by default).

priyanshu16095 replied:
It says that destinationDatabaseName has not been initialized, so I assigned it to an empty string ("").

subhramit commented on src/main/java/org/jabref/gui/maintable/RightClickMenu.java:
```diff
@@ -103,6 +115,51 @@ public static ContextMenu create(BibEntryTableViewModel entry,
+                if (bibDatabaseContext.getDatabasePath().isPresent()) {
+                    destinationPath = bibDatabaseContext.getDatabasePath();
+                    String uniquePathName = FileUtil.getUniquePathFragment(stateManager.collectAllDatabasePaths(), destinationPath.get()).get();
+                    if (uniquePathName.equals(sourceDatabaseName)) {
+                        return;
```
As far as I understand, here you are simply exiting if source and destination databases are the same. I think we should show a notification to the user in this case, just for convenience? @HoussemNasri

subhramit replied:
Ah, forgot about this one. Minor UX decision.
@HoussemNasri if necessary, we can do a follow-up.
@priyanshu16095 miscellaneous etiquette (which I also learnt late) - please don't resolve comments with decisions pending. If you push changes, the window will automatically be marked "outdated", and on taking the next look, the reviewer can close the conversation themselves if satisfied. This doesn't apply to reviews that are very trivially resolved (where you're 100% sure that it's the singular solution) - you can then use resolve.

subhramit replied:
Thanks for clarifying, can ignore if it's being taken care of by not listing. Good work. Resolving.

subhramit reviewed (commented):
Looks good from my end. Just a few nitpicks. Houssem and others will take a more thorough look.

subhramit commented:
Also, @priyanshu16095, please add a changelog entry.

priyanshu16095 commented:
@subhramit @Siedlerchr I mistakenly clicked the re-request review button. Please ignore.

Siedlerchr commented:
I took the liberty to test and improve the code a little bit with streams. works fine so far
